lib/constants.go

```diff
@@ -48,14 +47,53 @@ func EndTimer(timer *monitor.TimeMeasure) {
 	}
 }
 
+// WaitGroupWithError is like a sync.WaitGroup, with an error channel
+type WaitGroupWithError struct {
+	waiter  chan error
+	counter uint
+}
+
+// NewWaitGroupWithError creates a new WaitGroupWithError for the given count
+func NewWaitGroupWithError(count uint) WaitGroupWithError {
+	return WaitGroupWithError{
+		waiter:  make(chan error, count),
+		counter: count,
+	}
+}
+
+// Done mark the end of a goroutine, it has to be called, even with nil error
+func (wg WaitGroupWithError) Done(err error) {
+	wg.waiter <- err
+}
+
+// Wait waits for all expected goroutine to finish, returning the first error encountered
+func (wg WaitGroupWithError) Wait() error {
+	var ret error
+
+	for i := uint(0); i < wg.counter; i++ {
+		if err := <-wg.waiter; ret == nil && err != nil {
+			ret = err
+		}
+	}
+
+	return ret
+}
+
 // StartParallelize starts parallelization by instanciating number of threads
-func StartParallelize(nbrWg int) *sync.WaitGroup {
-	var wg sync.WaitGroup
-	wg.Add(nbrWg)
-	return &wg
+func StartParallelize(nbrWg uint) WaitGroupWithError {
+	return NewWaitGroupWithError(nbrWg)
+}
+
+// StartParallelizeWithInt starts parallelization by instanciating number of threads, panic if nbrWg < 0
+func StartParallelizeWithInt(nbrWg int) WaitGroupWithError {
+	if nbrWg < 0 {
+		panic("is negative")
```

Comment:
I know I proposed the `panic` - but now I saw it has an error channel. What makes more sense: panic or using the error channel?

Reply:
That's nicer to use the channel directly, bumping `StartParallelizeWithInt`.